File: test-client.py
#!/usr/bin/env python3
"""
Comprehensive test client for the routing server.

Usage:
    python test-client.py [server-ws-url]

Examples:
    python test-client.py                              # defaults to ws://localhost:3000
    python test-client.py wss://your-app.onrender.com

This script spawns TWO peers (A and B) and runs:
    1. Registration
    2. Heartbeat ping/pong
    3. Broadcast relay
    4. Targeted relay (A → B)
    5. Signal exchange (simulated WebRTC offer/answer)
    6. Peer list
    7. Disconnect notification
"""

import asyncio
import json
import sys
import os

# Force UTF-8 output on Windows consoles (emoji support)
if sys.platform == "win32":
    os.system("")  # Enable ANSI/VT100 escape sequences
    sys.stdout.reconfigure(encoding="utf-8", errors="replace")
    sys.stderr.reconfigure(encoding="utf-8", errors="replace")

try:
    import websockets
except ImportError:
    print("Missing dependency: websockets")
    print("Install it with:  pip install websockets")
    sys.exit(1)

SERVER = sys.argv[1] if len(sys.argv) > 1 else "ws://localhost:3000"
results: list[dict] = []


def log(tag: str, *args):
    print(f"[{tag}]", *args)


def pass_test(name: str):
    results.append({"name": name, "ok": True})
    print(f"  ✅ PASS: {name}")


def fail_test(name: str, reason: str):
    results.append({"name": name, "ok": False, "reason": reason})
    print(f"  ❌ FAIL: {name} — {reason}")


class Peer:
    """Wraps a WebSocket connection and provides message helpers."""

    def __init__(self, peer_id: str):
        self.id = peer_id
        self.ws = None
        self.inbox: list[dict] = []
        self.registered = False
        self.peer_id: str | None = None
        self._listeners: list[asyncio.Future] = []
        self._bg_task: asyncio.Task | None = None

    async def connect(self, timeout: float = 10.0):
        """Connect, register, and wait for the 'registered' response."""
        self.ws = await websockets.connect(SERVER)
        # Start background reader
        self._bg_task = asyncio.create_task(self._reader())
        # Send registration
        await self.ws.send(json.dumps({
            "type": "register",
            "peerId": self.id,
            "meta": {"label": self.id},
        }))
        # Wait for 'registered' message
        msg = await self.wait_for("registered", timeout=timeout)
        self.registered = True
        self.peer_id = msg["peerId"]
        return self

    async def _reader(self):
        """Background task that reads all incoming messages."""
        try:
            async for raw in self.ws:
                msg = json.loads(raw)
                self.inbox.append(msg)
                # Notify any waiters
                for fut in list(self._listeners):
                    if not fut.done():
                        fut.set_result(msg)
        except websockets.exceptions.ConnectionClosed:
            pass

    async def send(self, msg: dict):
        await self.ws.send(json.dumps(msg))

    async def send_and_wait(self, msg: dict, filter_type: str, timeout: float = 5.0) -> dict:
        """Send a message and wait for a specific response type."""
        wait_task = asyncio.ensure_future(self.wait_for(filter_type, timeout))
        await self.send(msg)
        return await wait_task

    async def wait_for(self, filter_type: str, timeout: float = 5.0) -> dict:
        """Wait for a message of a given type."""
        # Check inbox first (messages that already arrived)
        for msg in self.inbox:
            if msg["type"] == filter_type:
                self.inbox.remove(msg)
                return msg

        # Otherwise wait for future messages
        loop = asyncio.get_event_loop()
        while True:
            fut = loop.create_future()
            self._listeners.append(fut)
            try:
                msg = await asyncio.wait_for(fut, timeout=timeout)
                if msg["type"] == filter_type:
                    return msg
                # Not the type we want — keep waiting (reduce remaining timeout slightly)
            except asyncio.TimeoutError:
                raise TimeoutError(f'Timed out waiting for "{filter_type}"')
            finally:
                if fut in self._listeners:
                    self._listeners.remove(fut)

    async def close(self):
        if self._bg_task:
            self._bg_task.cancel()
        if self.ws:
            await self.ws.close()


async def run():
    print(f"\n🔌 Connecting to {SERVER}\n")

    # ── 1. Registration ─────────────────────────────────
    print("── TEST 1: Registration ──")
    peer_a = Peer("test-peer-A")
    peer_b = Peer("test-peer-B")

    try:
        await peer_a.connect()
        pass_test("Peer A registered")
    except Exception as e:
        fail_test("Peer A registration", str(e))
        sys.exit(1)

    try:
        await peer_b.connect()
        pass_test("Peer B registered")
    except Exception as e:
        fail_test("Peer B registration", str(e))
        await peer_a.close()
        sys.exit(1)

    # ── 2. Heartbeat ────────────────────────────────────
    print("\n── TEST 2: Heartbeat ping/pong ──")
    try:
        pong = await peer_a.send_and_wait({"type": "ping"}, "pong")
        if pong.get("ts") and isinstance(pong["ts"], (int, float)):
            pass_test("Heartbeat ping/pong")
        else:
            fail_test("Heartbeat ping/pong", "Missing or invalid ts in pong")
    except Exception as e:
        fail_test("Heartbeat ping/pong", str(e))

    # ── 3. Broadcast relay ──────────────────────────────
    print("\n── TEST 3: Broadcast relay ──")
    try:
        relay_task = asyncio.ensure_future(peer_b.wait_for("relay"))
        await peer_a.send({
            "type": "relay",
            "payload": {"text": "hello broadcast from A"},
        })
        relayed = await asyncio.wait_for(relay_task, timeout=5.0)
        if (relayed.get("from") == "test-peer-A"
                and relayed.get("payload", {}).get("text") == "hello broadcast from A"):
            pass_test("Broadcast relay A → B")
        else:
            fail_test("Broadcast relay A → B", "Unexpected relay content")
    except Exception as e:
        fail_test("Broadcast relay", str(e))

    # ── 4. Targeted relay ───────────────────────────────
    print("\n── TEST 4: Targeted relay (A → B) ──")
    try:
        # Drain any stale messages from previous tests
        await asyncio.sleep(0.5)
        peer_b.inbox = [m for m in peer_b.inbox if m["type"] != "relay"]
        relay_task = asyncio.ensure_future(peer_b.wait_for("relay"))
        await peer_a.send({
            "type": "relay",
            "to": "test-peer-B",
            "payload": {"secret": 42},
        })
        relayed = await asyncio.wait_for(relay_task, timeout=5.0)
        if (relayed.get("from") == "test-peer-A"
                and relayed.get("payload", {}).get("secret") == 42):
            pass_test("Targeted relay A → B")
        else:
            fail_test("Targeted relay A → B", "Unexpected payload")
    except Exception as e:
        fail_test("Targeted relay A → B", str(e))

    # ── 5. Signal exchange (UDP hole-punch simulation) ──
    print("\n── TEST 5: Signal exchange (hole-punch) ──")
    try:
        sig_task = asyncio.ensure_future(peer_b.wait_for("signal"))
        await peer_a.send({
            "type": "signal",
            "to": "test-peer-B",
            "signal": {"type": "offer", "sdp": "fake-sdp-offer"},
        })
        sig = await asyncio.wait_for(sig_task, timeout=5.0)
        if (sig.get("from") == "test-peer-A"
                and sig.get("signal", {}).get("type") == "offer"):
            pass_test("Signal offer A → B")
        else:
            fail_test("Signal offer A → B", "Unexpected signal content")

        # Answer back
        ans_task = asyncio.ensure_future(peer_a.wait_for("signal"))
        await peer_b.send({
            "type": "signal",
            "to": "test-peer-A",
            "signal": {"type": "answer", "sdp": "fake-sdp-answer"},
        })
        ans = await asyncio.wait_for(ans_task, timeout=5.0)
        if (ans.get("from") == "test-peer-B"
                and ans.get("signal", {}).get("type") == "answer"):
            pass_test("Signal answer B → A")
        else:
            fail_test("Signal answer B → A", "Unexpected signal content")
    except Exception as e:
        fail_test("Signal exchange", str(e))

    # ── 6. Peer list ────────────────────────────────────
    print("\n── TEST 6: Peer list ──")
    try:
        peer_list = await peer_a.send_and_wait({"type": "list"}, "peers")
        peers = peer_list.get("peers", [])
        if isinstance(peers, list) and len(peers) >= 2:
            pass_test(f"Peer list ({len(peers)} peers)")
        else:
            fail_test("Peer list", f"Expected >= 2 peers, got {len(peers)}")
    except Exception as e:
        fail_test("Peer list", str(e))

    # ── 7. Disconnect notification ──────────────────────
    print("\n── TEST 7: Disconnect notification ──")
    try:
        left_task = asyncio.ensure_future(peer_a.wait_for("peer_left"))
        await peer_b.close()
        left = await asyncio.wait_for(left_task, timeout=5.0)
        if left.get("peerId") == "test-peer-B":
            pass_test("Peer B disconnect notified to A")
        else:
            fail_test("Disconnect notification", "Wrong peerId in peer_left")
    except Exception as e:
        fail_test("Disconnect notification", str(e))

    # Wait 30 seconds before closing connections
    print("\n⏳ Waiting 30 seconds... check your stats page now!")
    await asyncio.sleep(30)

    # ── Summary ─────────────────────────────────────────
    await peer_a.close()
    print("\n════════════════════════════════════")
    print("         TEST SUMMARY")
    print("════════════════════════════════════")
    passed = sum(1 for r in results if r["ok"])
    failed = sum(1 for r in results if not r["ok"])
    for r in results:
        reason = f' ({r["reason"]})' if r.get("reason") else ""
        icon = "✅" if r["ok"] else "❌"
        print(f"  {icon} {r['name']}{reason}")
    print(f"\n  {passed} passed, {failed} failed\n")
    sys.exit(1 if failed > 0 else 0)


if __name__ == "__main__":
    try:
        asyncio.run(run())
    except Exception as e:
        print(f"Fatal error: {e}")
        sys.exit(1)

File: test-client.js
#!/usr/bin/env node
/**
 * Comprehensive test client for the routing server.
 *
 * Usage:
 *   node test-client.js [server-ws-url]
 *
 * Examples:
 *   node test-client.js                          # defaults to ws://localhost:3000
 *   node test-client.js wss://your-app.onrender.com
 *
 * This script spawns TWO peers (A and B) and runs:
 *   1. Registration
 *   2. Heartbeat ping/pong
 *   3. Broadcast relay
 *   4. Targeted relay (A → B)
 *   5. Signal exchange (simulated WebRTC offer/answer)
 *   6. Peer list
 *   7. Disconnect notification
 */

const WebSocket = require('ws');

const SERVER = process.argv[2] || 'ws://localhost:3000';
const results = [];

function log(tag, ...args) {
  console.log(`[${tag}]`, ...args);
}

function pass(name) {
  results.push({ name, ok: true });
  console.log(`  ✅ PASS: ${name}`);
}

function fail(name, reason) {
  results.push({ name, ok: false, reason });
  console.error(`  ❌ FAIL: ${name} — ${reason}`);
}

function createPeer(id) {
  return new Promise((resolve, reject) => {
    const ws = new WebSocket(SERVER);
    const peer = { id, ws, inbox: [], registered: false, peerId: null };

    ws.on('open', () => {
      ws.send(JSON.stringify({ type: 'register', peerId: id, meta: { label: id } }));
    });

    ws.on('message', (raw) => {
      const msg = JSON.parse(raw);
      peer.inbox.push(msg);
      if (msg.type === 'registered') {
        peer.registered = true;
        peer.peerId = msg.peerId;
        resolve(peer);
      }
    });

    ws.on('error', (e) => reject(e));

    // Timeout registration
    setTimeout(() => {
      if (!peer.registered) reject(new Error(`Peer ${id} registration timed out`));
    }, 10_000);
  });
}

function sendAndWait(peer, msg, filterType, timeoutMs = 5000) {
  return new Promise((resolve, reject) => {
    const handler = (raw) => {
      const m = JSON.parse(raw);
      if (m.type === filterType) {
        peer.ws.off('message', handler);
        resolve(m);
      }
    };
    peer.ws.on('message', handler);
    peer.ws.send(JSON.stringify(msg));
    setTimeout(() => {
      peer.ws.off('message', handler);
      reject(new Error(`Timed out waiting for "${filterType}"`));
    }, timeoutMs);
  });
}

function waitForMessage(peer, filterType, timeoutMs = 5000) {
  return new Promise((resolve, reject) => {
    const handler = (raw) => {
      const m = JSON.parse(raw);
      if (m.type === filterType) {
        peer.ws.off('message', handler);
        resolve(m);
      }
    };
    peer.ws.on('message', handler);
    setTimeout(() => {
      peer.ws.off('message', handler);
      reject(new Error(`Timed out waiting for "${filterType}"`));
    }, timeoutMs);
  });
}

async function run() {
  console.log(`\n🔌 Connecting to ${SERVER}\n`);

  // ── 1. Registration ─────────────────────────────────
  console.log('── TEST 1: Registration ──');
  let peerA, peerB;
  try {
    peerA = await createPeer('test-peer-A');
    pass('Peer A registered');
  } catch (e) {
    fail('Peer A registration', e.message);
    process.exit(1);
  }

  try {
    peerB = await createPeer('test-peer-B');
    pass('Peer B registered');
  } catch (e) {
    fail('Peer B registration', e.message);
    peerA.ws.close();
    process.exit(1);
  }

  // ── 2. Heartbeat ────────────────────────────────────
  console.log('\n── TEST 2: Heartbeat ping/pong ──');
  try {
    const pong = await sendAndWait(peerA, { type: 'ping' }, 'pong');
    if (pong.ts && typeof pong.ts === 'number') {
      pass('Heartbeat ping/pong');
    } else {
      fail('Heartbeat ping/pong', 'Missing or invalid ts in pong');
    }
  } catch (e) {
    fail('Heartbeat ping/pong', e.message);
  }

  // ── 3. Broadcast relay ──────────────────────────────
  console.log('\n── TEST 3: Broadcast relay ──');
  try {
    const relayPromise = waitForMessage(peerB, 'relay');
    peerA.ws.send(JSON.stringify({
      type: 'relay',
      payload: { text: 'hello broadcast from A' },
    }));
    const relayed = await relayPromise;
    if (relayed.from === 'test-peer-A' && relayed.payload?.text === 'hello broadcast from A') {
      pass('Broadcast relay A → B');
    } else {
      fail('Broadcast relay A → B', 'Unexpected relay content');
    }
  } catch (e) {
    fail('Broadcast relay', e.message);
  }

  // ── 4. Targeted relay ───────────────────────────────
  console.log('\n── TEST 4: Targeted relay (A → B) ──');
  try {
    const relayPromise = waitForMessage(peerB, 'relay');
    peerA.ws.send(JSON.stringify({
      type: 'relay',
      to: 'test-peer-B',
      payload: { secret: 42 },
    }));
    const relayed = await relayPromise;
    if (relayed.from === 'test-peer-A' && relayed.payload?.secret === 42) {
      pass('Targeted relay A → B');
    } else {
      fail('Targeted relay A → B', 'Unexpected payload');
    }
  } catch (e) {
    fail('Targeted relay A → B', e.message);
  }

  // ── 5. Signal exchange (UDP hole-punch simulation) ──
  console.log('\n── TEST 5: Signal exchange (hole-punch) ──');
  try {
    const sigPromise = waitForMessage(peerB, 'signal');
    peerA.ws.send(JSON.stringify({
      type: 'signal',
      to: 'test-peer-B',
      signal: { type: 'offer', sdp: 'fake-sdp-offer' },
    }));
    const sig = await sigPromise;
    if (sig.from === 'test-peer-A' && sig.signal?.type === 'offer') {
      pass('Signal offer A → B');
    } else {
      fail('Signal offer A → B', 'Unexpected signal content');
    }

    // Answer back
    const ansPromise = waitForMessage(peerA, 'signal');
    peerB.ws.send(JSON.stringify({
      type: 'signal',
      to: 'test-peer-A',
      signal: { type: 'answer', sdp: 'fake-sdp-answer' },
    }));
    const ans = await ansPromise;
    if (ans.from === 'test-peer-B' && ans.signal?.type === 'answer') {
      pass('Signal answer B → A');
    } else {
      fail('Signal answer B → A', 'Unexpected signal content');
    }
  } catch (e) {
    fail('Signal exchange', e.message);
  }

  // ── 6. Peer list ────────────────────────────────────
  console.log('\n── TEST 6: Peer list ──');
  try {
    const list = await sendAndWait(peerA, { type: 'list' }, 'peers');
    if (Array.isArray(list.peers) && list.peers.length >= 2) {
      pass(`Peer list (${list.peers.length} peers)`);
    } else {
      fail('Peer list', `Expected >= 2 peers, got ${list.peers?.length}`);
    }
  } catch (e) {
    fail('Peer list', e.message);
  }

  // ── 7. Disconnect notification ──────────────────────
  console.log('\n── TEST 7: Disconnect notification ──');
  try {
    const leftPromise = waitForMessage(peerA, 'peer_left');
    peerB.ws.close();
    const left = await leftPromise;
    if (left.peerId === 'test-peer-B') {
      pass('Peer B disconnect notified to A');
    } else {
      fail('Disconnect notification', 'Wrong peerId in peer_left');
    }
  } catch (e) {
    fail('Disconnect notification', e.message);
  }

  // ── Summary ─────────────────────────────────────────
  peerA.ws.close();
  console.log('\n════════════════════════════════════');
  console.log('         TEST SUMMARY');
  console.log('════════════════════════════════════');
  const passed = results.filter((r) => r.ok).length;
  const failed = results.filter((r) => !r.ok).length;
  results.forEach((r) => {
    console.log(`  ${r.ok ? '✅' : '❌'} ${r.name}${r.reason ? ` (${r.reason})` : ''}`);
  });
  console.log(`\n  ${passed} passed, ${failed} failed\n`);
  process.exit(failed > 0 ? 1 : 0);
}

run().catch((e) => {
  console.error('Fatal error:', e);
  process.exit(1);
});
